---
title: "Bioinformatics services - exercises"
author: "Julia Romanowska"
date: "`r Sys.Date()`"
output:
  prettydoc::html_pretty:
    theme: tactile
    highlight: github
    toc: true
---

```{r setup, include=FALSE}
library(readr)
library(dplyr)
library(tibble)
library(qvalue)
```

# Read the data

This was done for you, since the entire dataset is too big.

```{r, eval=FALSE}
diff_methyl <- read_csv("diffMethTable_site_cmp13.csv")
diff_methyl %>% glimpse()

qvals <- qvalue(diff_methyl$diffmeth.p.val)

diff_methyl <- diff_methyl %>%
  add_column(qval = qvals$qvalues)

# setting here to very low just to get not too many results
signif_qval <- 0.00001
diff_methyl_signif <- diff_methyl %>%
  filter(qval < signif_qval)
```

You have access to the file that contains only the most significant results

```{r}
diff_methyl_signif <- read_csv("diffMethTable.csv")
```

Now, `diff_methyl_signif` contains only the selected data.

# Write out results

Some services accept user data, but they require it to be in specific format.
It's easy to write various formats with R.

```{r}
# write out BED file
write_delim(
  diff_methyl_signif %>%
    mutate(End = Start + 1) %>%
    select(Chromosome, Start, End, Name = cgid, Score = diffmeth.p.val, Strand),
  path = "signif_CpGs.BED",
  delim = " ",
  col_names = FALSE
)

# write out region data
write_delim(
  diff_methyl_signif %>%
    mutate(strand_number = ifelse(Strand == "-", -1, 1)) %>%
    mutate(Chr = substr(Chromosome, 4, nchar(Chromosome))) %>%
    mutate(reg = paste0(Chr, ":", Start, ":", Start + 1, ":", strand_number)) %>%
    select(reg),
  path = "signif_CpGs.regions",
  delim = " ",
  col_names = FALSE
)

write_delim(
  diff_methyl_signif %>%
    mutate(Chr = substr(Chromosome, 4, nchar(Chromosome))) %>%
    mutate(reg = paste0(Chr, ":", Start, ":", Start + 1)) %>%
    select(reg),
  path = "signif_CpGs_noStrand.regions",
  delim = " ",
  col_names = FALSE
)
```

----

# Web services

We will take the results above and perform several searches online.

## ensembl BioMart - genes

Go to [ensembl](http://www.ensembl.org/biomart/martview/) and use `signif_CpGs.regions`
file to extract the genes that lay near CpGs.

- Choose database: Ensembl Genes
- Choose dataset: Human Genes
- On the left (blue menu), choose `Filters`, expand `REGION` and select `Multiple regions`, then Browse to the file
- Next, choose `Attributes` (below `Filters`) and expand `GENE`, then choose the following:

    - Gene stable ID
    - Protein stable ID
    - Gene description
    - Chromosome/scaffold name
    - Gene start (bp)
    - Gene end (bp)
    - Gene name

- Click on `Results` (above `Filters`) and choose to "Export all results to | File | TSV", tick off "Unique results only" and click "Go"
- Save the file to the disk, we will read it below

```{r}
# after getting results from ensembl BioMart
genes_near_signif_cpgs <- read_delim(file = "mart_export.txt", delim = "\t",
  col_names = c("Gene_ID", "Protein_ID", "Chr", "Start", "End", "Gene_name", "Gene_descr"),
  skip = 1)
genes_near_signif_cpgs

knitr::kable(genes_near_signif_cpgs %>%
  count(Chr))
```

To use the results in the next step, we need only a list of the gene names.

```{r}
write_delim(genes_near_signif_cpgs %>%
              distinct(Gene_name),
            path = "genes_near_signif_cpgs.txt", delim = " ", col_names = FALSE)
```

## STRING - database of protein-protein interactions (PPI)

We can check whether our set of genes is enriched in something specific or has
other common features or functions by e.g., checking it in [STRING](https://string-db.org/)

- Click on SEARCH
- Choose `Multiple proteins` in the list to the left
- Click on the "Browse" button and navigate to `genes_near_signif_cpgs.txt` file
- Enter `Homo sapiens` in the "Organism" field and click "Search"
- You will be presented with possible matches to the identifiers that we provided - accept all and continue
- Explore the network, e.g.,

    - in the "Analysis" tab, check enrichment
    - in the "Clusters" tab, play with k-means clustering
    - what is the protein "in the middle"?

## ensembl BioMart - regulatory regions

ensembl browser has many databases - one contains regulatory features of the
genome (such as promoters and enhancers). Export the regulatory regions that
are in vicinity of our selected CpGs.

- Choose database: Ensembl Regulation
- Choose dataset: Human Regulatory Features
- On the left (blue menu), choose `Filters`, expand `REGION` and select `Multiple regions`, then Browse to the file `signif_CpGs_noStrand.regions`
- Next, choose `Attributes` (below `Filters`) and expand `GENE`, then choose the following:

    - Chromosome/scaffold name
    - Start (bp)
    - End (bp)
    - Feature type
    - Regulatory stable ID
    - SO term accession

- Click on `Results` (above `Filters`) and choose to "Export all results to | File | TSV", tick off "Unique results only" and click "Go"
- Save the file to the disk as `mart_export_regulatory_feat.txt`, we will read it below

```{r}
# after checking regulatory regions
regulatory_near_signif_cpgs <- read_delim(
  file = "mart_export_regulatory_feat.txt",
  delim = "\t",
  skip = 1,
  col_names = c("chr", "start", "end", "feat_type", "reg_ID", "SO_term")
)
regulatory_near_signif_cpgs
regulatory_near_signif_cpgs %>%
  count(SO_term) %>%
  left_join(regulatory_near_signif_cpgs %>%
              distinct(SO_term, feat_type))

regulatory_near_signif_cpgs %>%
  filter(feat_type == "Enhancer")
```

How many different regulatory features were found?

## GenEnhancer by GeneCards

Choose one enhancer (by `reg_ID` - this is the ensembl ID of the region) and
enter it in the "Keywords" search field at https://www.genecards.org/. _(NOTE:
this is not the main search field, which only searches for gene names)_

What genes were found? Check their Cards and see where are they located
on the genome in relation to the enhancer you've chosen.
