---
title: "NORBIS GENESTAT course, R intro"
author: "Ã˜ystein Haaland, Miriam Gjerdevik, Julia Romanowska"
date: "`r Sys.Date()`"
output:
  prettydoc::html_pretty:
    theme: tactile
    highlight: github
    toc: true
---

# PART 1

> Note: throughout the course, we will mix between the `base` R notation and `tidyverse` notation because it is not always the easiest to use the `tidyverse` way on these genetic data.

```{r read_libs}
library(dplyr)
library(tidyr)
library(readr)
library(snpStats)
```

## SNPs and alleles

1. Read data from file `data_a_intro_R.dat` and save it to `obj`:

```{r read_input_data}
obj <- read_table2("data_a_intro_R.dat")
```

2. Look at the data: check the class, dimensions and print the first part of the data:

```{r}
class(obj)
obj
```

3. Using the data

  + Accessing the column named `cc.status`, using the `$` sign (0 codes for control and 1 for case)

```{r}
cc.status <- obj$cc.status
class(cc.status)
mean(cc.status)
summary(cc.status)
table(cc.status)
```

  + Accessing the first SNP
  
```{r}
snp1 <- obj %>%
  select(snp_1.1, snp_1.2)
snp1
```

  + Tabulating the first SNP

```{r}
table(snp1$snp_1.1)
table(snp1$snp_1.2)
table(snp1)
```

  + Divide by number of indivdiuals (function `nrow` will count all the rows, i.e., individuals) and check whether it sums up to 100%

```{r}
table(snp1)/nrow(snp1)
sum(table(snp1)/nrow(snp1))
```

  + Rounding to two decimal points - still sums up to 100%!

```{r}
round(table(snp1)/nrow(snp1), 2)
sum(round(table(snp1)/nrow(snp1), 2))
```

  + Tabulate SNP8
  
```{r}
snp8 <- obj %>%
  select(snp_8.1, snp_8.2)
snp8
table(snp8$snp_8.1)
table(snp8$snp_8.2)
table(snp8)
```

  + `-9` means missing. Replace `-9` with `NA` (not available) in the entire dataset.

```{r}
obj[obj == -9] <- NA
snp8 <- obj %>%
  select(snp_8.1, snp_8.2)
```

### Exercise 1

Fill the space between apostrophes with your code and click the green arrow to check how it evaluates.

1.1 Tabulate SNP8 again.
```{r ex-1-1}

```

1.2 What is the distribution of alleles?
```{r ex-1-2}

```

1.3 Why does it not sum to 100%? Hint: `?table`

```{r ex-1-3}

```

---

## Missing data, apply, which, names, any, remove rows and columns from object

1. Set the argument `useNA = "ifany"` to make the distribution of alleles in SNP8 sum up to 100%

```{r}
table(snp8, useNA = "ifany")/nrow(snp8)
sum(table(snp8, useNA = "ifany")/nrow(snp8))
```

2. Find those columns in `obj` that contain missing data

```{r}
na.obj <- is.na(obj)
head(na.obj)
head(na.obj)*1
```

  + Function `is.na` returns `TRUE` or `FALSE` for each value, but we can translate it to 0's and 1's, which is easier to handle when calculating some properties of each SNP.

3. Identify proportion of missing data for each SNP

```{r}
na.snp <- apply(na.obj, 2, mean)
na.snp
```

  + Which SNPs have more than 7% missing?

```{r}
na.snp.07 <- which(na.snp > 0.07)
na.snp.07
names(na.snp.07)
```

  + Look at first 10 rows of SNPs with more than 7% missing

```{r}
obj %>% select(!!names(na.snp.07))
```

  + Which SNPs have any missing data?

```{r}
apply(na.obj, 2, any)
```

We can do the same in two different ways: 

* directly on the original data set:

```{r}
apply(is.na(obj), 2, any)
```

* and an advanced version, using a self-written function (useful if you must create your own code that is executed on all columns):

```{r}
apply(obj, 2, function(x){
  any(is.na(x))
})
```

  + Remove SNPs with more than 7% missing:

```{r}
obj <- obj %>%
  select(-!!names(na.snp.07))
obj
```

### Exercise 2

2.1 Which individuals have any missing data? (Use `apply` and `obj`, not `na.obj`)

```{r ex-2-1}

```

```{r ex-2-1-hint, eval=FALSE}
which(apply(obj, 1, function(x){
	...
}))
```

2.2 Keep individuals with less than 10% missing. Hint: First identify the rows with more than 10% missing, and then remove those.
```{r ex-2-2}

```

```{r ex-2-2-hint, eval=FALSE}
more10.missing <- which(apply(obj, 1, function(x){
	sum(is.na(x)) > length(x)/10
}))
obj.less10missing <- ...
obj.less10missing
```

---

## Merging matrices

1. Start with the original dataset

```{r}
obj <- read_table2("data_a_intro_R.dat")
```

2. `data_b_intro_R.dat` contains data on SNP24. Combine the previous dataset and this new SNP (`bind_cols` combines _columnwise_).

```{r}
snp24 <- read_table2("data_b_intro_R.dat")
snp24
obj <- obj %>%
  bind_cols(snp24)
obj
```

### Exercise 3

3.1 Add extra individuals (rows!) from `data_c_intro_R.dat` (with function `bind_rows`).

```{r ex-3-1}

```

---

## Save and load objects

- `write_table` writes a data.frame, tibble, or matrix into a text-formatted file.
- `save` saves objects (not only matrices) to a binary file, readable by R with the `load` function.
- `ls` function shows what objects are currently loaded into memory.

```{r}
write_delim(obj, path = "data_prep_R.dat", col_names = TRUE)

save(obj, file = "obj.RData")
rm(obj)
ls()
load("obj.RData")
ls()
```

---

## Packages

One of the strengths of R is the vast number of packages that can be installed at no cost. 

1. Installing package(s)
  
```{r, eval=FALSE}
install.packages("Haplin)
```

2. Loading a package

```{r}
library(Haplin)
```

3. Help function for entire package

```{r}
help(package = "Haplin")
```

---

## Save and load workspace.

This saves the entire content of the memory, so that one can get back to work where they'd finished. Note that this does not re-load libraries, so one needs to run all the necessary `library` commands before or after loading a new workspace.

```{r, eval=FALSE}
save.image("genestat.RData")

load("genestat.RData")
```

----

# PART 2

We are now starting to work on real genotype data!

PLINK (Will get back to this in other lectures)

Input files: `data.bed`, `data.bim`, `data.fam`. _Not available to you!_
`plink --bfile data --alleleACGT --recode --out data`
The output is found in `pres.ped` and `pres.map`. _Available to you!_

The files `data.ped` and `data.map` contain information about the genotype and phenotype, and about the SNPs, respectively, for a dataset with families (mother, father and a child), where the child had an oral cleft. 

1. Show the first 10 lines of `data.map`
  
```{r}
read_table2("data.map", col_names = FALSE)
```

2. Show the first 10 lines of `data.ped`

```{r}
read_table2("data.ped", col_names = FALSE)
```

PED-format:

FAMILY_ID | ID_CHILD | ID_DAD | ID_MOM | SEX | CC | GENOTYPES 
----------|----------|--------|--------|-----|----|-----------
1         | 1_1      | 1_3    | 1_2    | 2   | 1  | G G T A A T C G C G G A ...
1         | 1_2      | 0      | 0      | 2   | 0  | G G T A A T G G G G A A ...
1         | 1_3      | 0      | 0      | 1   | 1  | G G T A A T C G C C G A ...
2         | 2_1      | 2_3    | 2_2    | 1   | 0  | G G T A A T G G C G G A ...
2         | 2_2      | 0      | 0      | 2   | 0  | G G A A T T G G G G A A ...
2         | 2_3      | 0      | 0      | 1   | 1  | 0 0 0 0 0 0 0 0 0 0 0 0 ...


## Reading a PED file with snpStats

Read ped-file into R

```{r}
raw.all <- read.pedfile(file = "data.ped", snps = "data.map", which = 2)
```

## Look at the data

1. Structure of the data

```{r}
str(raw.all)
```

2. Renaming columns to match what we have in the `.map` file

```{r}
colnames(raw.all$map)[c(1,3)] <- c("chromosome", "position")
head(raw.all$map)
```

3. Renaming rows to match what we have in the `.map` file

```{r}
rownames(raw.all$map)
rownames(raw.all$map) <- raw.all$map$snp.names
head(raw.all$map)
```

4. Look at structure again

```{r}
str(raw.all)
```

5. Phenotype data (`fam`) contains information about individuals: ID, sex, case or control. NOTE: snpStats codes `0` as `NA` in case/control status column, so we need to fix it.

```{r}
head(raw.all$fam)
raw.all$fam$affected[is.na(raw.all$fam$affected)] <- 0
head(raw.all$fam)
```

6. Check the number of individuals

```{r}
nrow(raw.all$fam)
```

7. Tabulate the individuals' gender

```{r}
table(raw.all$fam$sex)
```

8. Get the chromosome names

```{r}
unique(raw.all$map$chromosome)
```

### Exercise 4

4.1 Tabulate the chromosome names.

```{r ex-4-1}

```

4.2 What are the dimensions of the genotype data? Are the individuals represented by columns or rows? (Hint: check the structure of `raw.all`)

```{r ex-4-2}

```

4.3 Check how many _families_ are in the data?

**Hint:** tabulate the `pedigree` column of the family information

```{r ex-4-3}

```

4.4 How many of the individuals were affected?

```{r ex-4-4}

```

---

## Extract children, chromosome and SNPs

1. Find all the individuals with an ID ending with `_1` (these are children).

```{r}
head(raw.all$fam$member, 21)
children <- grep(pattern = "_1", x = raw.all$fam$member)
head(children, 30)
head(raw.all$fam$member[children], 30)
```

2. Make a new object with children only

```{r}
raw.child.gen <- raw.all$genotypes[children,]
str(raw.child.gen)
raw.child.fam <- raw.all$fam[children,]
raw.child <- list(genotypes = raw.child.gen, 
                  fam = raw.child.fam, 
                  map = raw.all$map)
str(raw.child)
```

3. Make a new object with chromosome 1 only

* check which SNPs are in chromosome 1

```{r}
chr1.snps <- raw.all$map$snp.names[raw.all$map$chromosome == "1"]
chr1.snps
```

* since this is a factor, we must reformat it to numeric, to use it in selection

```{r}
chr1.snps <- as.character(levels(chr1.snps)[chr1.snps])
```

* take only those SNPs from the genotype matrix

```{r}
raw.child$genotypes <- raw.child$genotypes[, chr1.snps]
raw.child$map <- raw.child$map[chr1.snps ,]
str(raw.child)
```

4. Make a new object with first ten SNPs only

```{r}
raw.child.tmp <- raw.child$genotypes[, 1:10]
str(raw.child.tmp)
```

5. Make a new object with the following SNPs: rs12, rs90, rs93, rs107

```{r}
head(colnames(raw.child$genotypes))
raw.child.tmp <- raw.child$genotypes[,
  colnames(raw.child$genotypes) %in% c("rs12", "rs90", "rs93", "rs107")]
str(raw.child.tmp)
```

### Exercise 5

5.1 How many _children_ were affected?

```{r ex-5-1}
```

5.2 Create an object `raw.mothers` with mothers only

```{r ex-5-2}
```

```{r ex-5-2-hint, eval=FALSE}
mothers <- grep(pattern = "_2", x = raw.all$fam$member)
```

5.3 From this new matrix, remove the following SNPs: rs12, rs90

**Hint:** exclamation mark `!` gives a negation of a boolean operator

```{r ex-5-3}
```

---

## Quality control

1. Summary of the genetic data, per individual:

```{r}
qc.child <- row.summary(raw.child.gen)
summary(qc.child)

plot(qc.child)
```

2. We don't want the individuals with a call rate lower than `0.95`:

```{r}
ok.call.rate <- qc.child$Call.rate >= 0.95
raw.child.ok <- list(genotypes = raw.child$genotypes[ok.call.rate,],
                      fam = raw.child$fam[ok.call.rate,],
                      map = raw.child$map)
str(raw.child.ok)
```

3. Summary of the genetic data, per marker:

```{r}
qc.child.marker <- col.summary(raw.child.ok$genotypes)
summary(qc.child.marker)
```

4. We don't want the markers with a MAF lower than `0.05`:

```{r}
ok.MAF <- qc.child.marker$MAF >= 0.05
raw.child.ok <- list(genotypes = raw.child.ok$genotypes[,ok.MAF],
                      fam = raw.child.ok$fam,
                      map = raw.child.ok$map[ok.MAF,])
str(raw.child.ok)
```

### Exercise 6

6.1 Remove the _markers_ with a call rate lower than `0.99`:

```{r ex-6-1}
```

6.2 How many individuals and markers remain?

```{r ex-6-2}
```

----

# PART 3

## Association analyses in snpStats

1. Prepare the "clean" data for `single.snp.tests()`.

```{r}
support.data <- data.frame(cc = raw.child.ok$fam$affected)
rownames(support.data) <- rownames(raw.child.ok$fam)
head(support.data)
```

2. Run the analysis.

```{r}
gwa.child <- single.snp.tests(phenotype = cc, 
                               data = support.data, 
                               snp.data = raw.child.ok$genotypes)
summary(gwa.child)
```

3. Make QQ-plot.

```{r}
chi2 <- chi.squared(gwa.child, df = 1)
qq.chisq(chi2, df = 1)
```

4. Retrieve p-values.

```{r}
pval <- p.value(gwa.child, df = 1)
head(sort(pval), 20)
```

5. Find the most signif.SNPs

```{r}
ord <- order(pval)
top10 <- head(ord, 10)
top10

top10.names <- gwa.child@snp.names[top10]
top10.names
```

### Exercise 7

7.1 Repeat steps 3 to 5 using the p-values for the test with 2 df. Store the new p-values in pval2.

```{r ex-7-1}
```


```{r ex-7-1-hint, eval=FALSE}
chi2.2df <- chi.squared(gwa.child, df = 2)

```

7.2 How many SNPs were in the top 10 using both methods? 

```{r ex-7-2}
```

7.3 Plot pval vs. pval2 

```{r ex-7-3}
```

7.4 Use `abline` to draw a straight line through the plot with `intercept = 0` and `slope = 1`. How would you interpret this line? 

```{r ex-7-4}
```


